components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/CORSRequestHandler.java

```diff
@@ -16,69 +16,191 @@
 
-    private static final Log log = LogFactory.getLog(CORSRequestHandler.class);
-
-    public void init(SynapseEnvironment synapseEnvironment) {
-        log.debug("Initializing API authentication handler instance");
-    }
-
-    public void destroy() {
-        log.debug("Destroying API authentication handler instance");
-    }
-
-    public boolean handleRequest(MessageContext messageContext) {
-
-        if (Utils.isCORSEnabled()) {
-	    	/* For CORS support adding required headers to the response */
-            org.apache.axis2.context.MessageContext axis2MC = ((Axis2MessageContext) messageContext).
-                    getAxis2MessageContext();
-            Map<String, String> headers = (Map) axis2MC.getProperty(org.apache.axis2.context.MessageContext.TRANSPORT_HEADERS);
-            String requestOrigin = (String) headers.get("Origin");
-
-            headers.put(APIConstants.CORSHeaders.ACCESS_CONTROL_ALLOW_ORIGIN, Utils.getAllowedOrigin(requestOrigin));
-            headers.put(APIConstants.CORSHeaders.ACCESS_CONTROL_ALLOW_METHODS, Utils.getAllowedMethods());
-            headers.put(APIConstants.CORSHeaders.ACCESS_CONTROL_ALLOW_HEADERS, Utils.getAllowedHeaders());
-            axis2MC.setProperty(org.apache.axis2.context.MessageContext.TRANSPORT_HEADERS, headers);
-        }
-
-        return true;
-    }
-
-    public boolean handleResponse(MessageContext messageContext) {
-
-    	if (Utils.isCORSEnabled()) {
-	    	/* For CORS support adding required headers to the response */
-	    	org.apache.axis2.context.MessageContext axis2MC = ((Axis2MessageContext) messageContext).
-	                getAxis2MessageContext();
-	    	Map<String, String> headers = (Map) axis2MC.getProperty(org.apache.axis2.context.MessageContext.TRANSPORT_HEADERS);
-            String requestOrigin = (String) headers.get("Origin");
-	    		    	    	
-	    	headers.put(APIConstants.CORSHeaders.ACCESS_CONTROL_ALLOW_ORIGIN, Utils.getAllowedOrigin(requestOrigin));
-	        headers.put(APIConstants.CORSHeaders.ACCESS_CONTROL_ALLOW_METHODS, Utils.getAllowedMethods());
-	        headers.put(APIConstants.CORSHeaders.ACCESS_CONTROL_ALLOW_HEADERS, Utils.getAllowedHeaders());
-	        axis2MC.setProperty(org.apache.axis2.context.MessageContext.TRANSPORT_HEADERS, headers);
-    	}
- 
-        return true;
-    }
+	private static final Log log = LogFactory.getLog(CORSRequestHandler.class);
+	private String inline;
+	private String allowHeaders;
+	private List<String> allowedOrigins;
+	private boolean headerStatus;
+
+	public void init(SynapseEnvironment synapseEnvironment) {
+		if (log.isDebugEnabled()) {
+			log.debug("Initializing CORSRequest Handler instance");
+		}
+		if (ServiceReferenceHolder.getInstance().getApiManagerConfigurationService() != null) {
+			headerStatus = initializeHeaders();
+		}
+	}
+
+	public boolean initializeHeaders() {
+		if (allowHeaders == null) {
+			allowHeaders = Utils
+					.getAllowedHeaders();
```

Comment:
Max character limit for a line is 120. Are you sure you are using that in the code formatting template?